finality-aleph/src/metrics/finality_rate.rs

```diff
@@ -0,0 +1,162 @@
+use std::collections::{hash_map::Entry, HashMap};
+
+use primitives::{BlockHash, BlockNumber};
+use substrate_prometheus_endpoint::{register, Counter, PrometheusError, Registry, U64};
+
+#[derive(Clone)]
+pub enum FinalityRateMetrics {
+    Prometheus {
+        own_finalized: Counter<U64>,
+        own_hopeless: Counter<U64>,
+        imported_cache: HashMap<BlockNumber, Vec<BlockHash>>,
+    },
+    Noop,
+}
+
+impl FinalityRateMetrics {
+    pub fn new(registry: Option<&Registry>) -> Result<Self, PrometheusError> {
+        let registry = match registry {
+            None => return Ok(FinalityRateMetrics::Noop),
+            Some(registry) => registry,
+        };
+
+        Ok(FinalityRateMetrics::Prometheus {
+            own_finalized: register(
+                Counter::new("aleph_own_finalized_blocks", "no help")?,
+                registry,
+            )?,
+            own_hopeless: register(
+                Counter::new("aleph_own_hopeless_blocks", "no help")?,
+                registry,
+            )?,
+            imported_cache: HashMap::new(),
+        })
+    }
+
+    /// Stores the imported block's hash. Assumes that the imported block is own.
+    pub fn report_own_imported(&mut self, hash: BlockHash, number: BlockNumber) {
+        let imported_cache = match self {
+            FinalityRateMetrics::Prometheus { imported_cache, .. } => imported_cache,
+            FinalityRateMetrics::Noop => return,
+        };
+
+        match imported_cache.entry(number) {
+            Entry::Occupied(mut entry) => entry.get_mut().push(hash),
+            Entry::Vacant(entry) => {
+                entry.insert(vec![hash]);
+            }
+        }
+    }
+
+    /// Counts the blocks at the level of `number` different than the passed block
+    /// and reports them as hopeless. If `hash` is a hash of own block it will be found
+    /// in `imported_cache` and reported as finalized.
+    pub fn report_finalized(&mut self, hash: BlockHash, number: BlockNumber) {
+        let (own_finalized, own_hopeless, imported_cache) = match self {
+            FinalityRateMetrics::Prometheus {
+                own_finalized,
+                own_hopeless,
+                imported_cache,
+            } => (own_finalized, own_hopeless, imported_cache),
+            FinalityRateMetrics::Noop => return,
+        };
+
+        match imported_cache.entry(number) {
+            Entry::Occupied(entry) => {
+                own_hopeless.inc_by(
+                    entry
+                        .get()
+                        .iter()
+                        .filter(|h| {
+                            if **h == hash {
+                                own_finalized.inc();
+                            }
+                            **h != hash
+                        })
+                        .count() as u64,
+                );
+                entry.remove();
```

Comment:
Not all blocks in the chain history have justifications. That means that fresh node will kepp in `imported_cache` all of them and will only get rid of them in case of restart. Lets introduce some form of pruning old entries in the cache.

Reply:
Maybe use LruCache and bound the vector inside to avoid attacks that would cause out of memory crashes (even if they are unpractical)?

Reply:
Hmm, the blocks we store in `imported_cache` are only the blocks we produced. If we produced a block on some `level` I assume that `top_finalized < level`. So for each key in this `imported_cache` eventually there should come a finalized block which will erase it. 

That's my reasoning, but still I'm eager to implement some pruning mechanism just in case...

Reply:
In `Vec` there are our produced blocks at a particular level. Given the recent precaution measures in terms of equivocations these `Vecs` should be at most singletons :stuck_out_tongue:

I'm thinking on a different mechanism - Given that our produced blocks come in order of increasing level, `imported_cache` could just be a `Deque` - when a block comes, push it back, when a finalized block comes, erase the front until the appropriate level. Just need to verify with our main `sync` guys that block do come in such order from sync.

Reply:
some combination of LruCache plus bounded vec looks ok

Reply:
Changed inner data type from `LruCache` to `BoundedVec`

Reply:
Ok, Im satisfied with this. @ggawryal if you dont have any more comments here, resolve this conversation.

Reply:
I forgot that there are only our blocks in the cache... In this case plain vec probably would be sufficient, giving a (theoretical) bound of the size of a block producing committee, but we can defensively leave it as it is :)